### ISSUE TITLE
About depth_distribution_similarity_loss

### ISSUE THREAD

**Issue by k73kondrashov** [doubleZ0108/GeoMVSNet#8]
Hi! Thank you for the good work!
I am trying to train the network on DTU using your scripts and I have doubts about depth_distribution_similarity_loss. 

pixel_wise_loss starts around 1.0 and decreases, dds_loss does not decrease and ranges from 140 to 80. At the same time, the predictions are still good

I looked at the code, you pass the depth values to kl-divergence, as far as I know kl-divergence usually expects the probability of distribution as inputs. And your article says that p and q are depth distributions

Can you explain why this is so?

P.S. I use batch_size = 1. I hope that's not the problem

```
def depth_distribution_similarity_loss(depth, depth_gt, mask, depth_min, depth_max):
    depth_norm = depth * 128 / (depth_max - depth_min)[:,None,None]
    depth_gt_norm = depth_gt * 128 / (depth_max - depth_min)[:,None,None]

    M_bins = 48
    kl_min = torch.min(torch.min(depth_gt), depth.mean()-3.*depth.std())
    kl_max = torch.max(torch.max(depth_gt), depth.mean()+3.*depth.std())
    bins = torch.linspace(kl_min, kl_max, steps=M_bins)

    kl_divs = []
    for i in range(len(bins) - 1):
        bin_mask = (depth_gt >= bins[i]) & (depth_gt < bins[i+1])
        merged_mask = mask & bin_mask 

        if merged_mask.sum() > 0:
            p = depth_norm[merged_mask] # is depth values, not probs
            q = depth_gt_norm[merged_mask]
            kl_div = torch.nn.functional.kl_div(torch.log(p)-torch.log(q), p, reduction='batchmean')
            kl_div = torch.log(kl_div)
            kl_divs.append(kl_div)

    dds_loss = sum(kl_divs)
    return dds_loss
```
![image](https://github.com/doubleZ0108/GeoMVSNet/assets/90474417/6520b1ab-6786-485f-ad0f-c0631c3f1a61)

**Comment by doubleZ0108** [comment]
Hi, thanks for your interest in our work.

Firstly, to be honest, `batch_size=1` always meets some "ghost" problem, so please check it first.

As for your problem, as described in Sec. 3.3 in our paper, we found that the data distribution satisfies the Mixed-Gaussian Model. So we pretend to calculate the similarity of two depth distributions (estimated depth & depth GT). That's the motivation of the code you referenced in our implementation.

As for
> pixel_wise_loss starts around 1.0 and decreases, dds_loss does not decrease and ranges from 140 to 80. At the same time, the predictions are still good

We need to admit that the data of small scenes such as DTU do not fully meet the perfect Gaussian model, so different training iterations can not accurately describe the similarity of curves, resulting in loss shaking in the training of this loss.

So a better way to characterize the depth of the whole scene (v.s. pixel-wise loss measure) will be one of the directions of future development, I think.

We are really sorry for confusing you and hope the reply helps.

**Comment by xingchen2022** [comment]
@doubleZ0108 hi, I also have some doubts about the code of dds_loss.

1) In Section 3.3 of the paper, it is mentioned that a GMM with K = 1 and K = 2 is used to fit the full-scene depth distribution. However, in the code, only one Gaussian component is utilized. Does this imply that using one Gaussian component to calculate the depth value range for representing the distribution of each scene is sufficient?

2) After obtaining the depth interval of 1 (μi - 3σi, μi + 3σi), why do you use 48 smaller bins to calculate KL divergence between the predicted depth and ground truth depth for each bin, rather than calculating the entire scene at once?

3) When normalizing the depth, you use the formula depth * 128 / (depth_max - depth_min). Why not use (depth - depth_min) * 128 / (depth_max - depth_min), which is a more commonly used normalizing method? 

4)I have used the script to train the network without making any parameter adjustments (batch size = 4 with 2 GPUs). 
![image](https://github.com/doubleZ0108/GeoMVSNet/assets/40599481/743a25f9-ca23-4142-b6bc-1fea6548a705)
I found the dds_loss curve in each training stage didn't exhibits a decrease trend. I'm curious whether this will lead to improved reconstruction results? (And the pw_loss of stage 0 has also increased as training continues)

These questions might be a bit extensive, but they've been causing me confusion for some time. I would greatly appreciate it if you could provide some clarifications. Thank you in advance!

**Comment by doubleZ0108** [comment]
@xingchen2022 Yes, dds_loss is not perfect. The mixed Gaussian model analyzed in Section 3.3 is our ultimate vision, but dds_loss is a trade-off. It does not perfectly decrease in all training iterations, but I believe that the loss part of characterizing the curve can still guide the network to some extent (just like the momentum strategy). As for the setting of 48, it is because we use binning for calculation, which involves continuous and discrete conversions.

**Comment by xingchen2022** [comment]
> @xingchen2022 Yes, dds_loss is not perfect. The mixed Gaussian model analyzed in Section 3.3 is our ultimate vision, but dds_loss is a trade-off. It does not perfectly decrease in all training iterations, but I believe that the loss part of characterizing the curve can still guide the network to some extent (just like the momentum strategy). As for the setting of 48, it is because we use binning for calculation, which involves continuous and discrete conversions.

Thank you very much, It's valuable information!  And could you tell me your insights regarding the depth normalizing code? (Q3: you use the formula depth * 128 / (depth_max - depth_min). Why not use (depth - depth_min) * 128 / (depth_max - depth_min), which is a more commonly used normalizing method?)

**Comment by doubleZ0108** [comment]
@xingchen2022 Sorry for the missing. Yeah, the way you said is correct and I have tried it during my experiments before. However, I found that the training effect was not as good as it is now. I think that normalization is just a strategy for data scaling and does not have strict mathematical constraints, so `- depth_min` is not necessary in my opinion.

**Comment by xingchen2022** [comment]
> @xingchen2022 Sorry for the missing. Yeah, the way you said is correct and I have tried during my experiments before. However, I found that the training effect was not as good as it is now. I think that normalization is just a strategy for data scaling and does not have strict mathematical constraints, so -depth_min is not necessary in my opinion.

Got it! Better result is all we want, hhh
Thanks again, and wish you have a good rest tonight ❤

**Comment by k73kondrashov** [comment]
@doubleZ0108 Thank you for your answers
I think the issue may be closed
